### PR TITLE
[bazel] compile capnp

### PR DESCRIPTION
**Pull request by mikea**
Provides cc_capnp_library for capnp compilation.

### CODE REVIEW

File: c++/src/capnp/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    ],
+)
+
+filegroup(
+    name = "testdata",
+    srcs = glob(["testdata/**/*"]),
+)
+
+cc_capnp_library(
+    name = "test_capnp",
```
Comment (uhthomas) [inline c++/src/capnp/BUILD.bazel:189]: the convention is capnp_test, or capnp_test_library I believe

File: c++/src/capnp/cc_capnp_library.bzl [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        includes.append(src_prefix)
+
+    system_include = _workspace_path(ctx.attr._capnp_system.label, "src")
+
+    out_dir = ctx.var["GENDIR"]
+    if src_prefix != "":
+        out_dir = out_dir + "/" + src_prefix
+
+    cc_out = "-o%s:%s" % (ctx.executable._capnpc_cxx.path, out_dir)
+    args = ["compile", "--verbose", cc_out] + \
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:34]: prefer ctx.actions.args()

File: c++/src/capnp/cc_capnp_library.bzl [hunk 3]
```diff
@@ -0,0 +1,114 @@
+"""Bazel rule to compile .capnp files into c++."""
+
+capnp_provider = provider("Capnproto Provider", fields = {
+    "includes": "includes for this target (transitive)",
+    "inputs": "src + data for the target",
+    "src_prefix": "src_prefix of the target",
+})
+
+def _workspace_path(label, path):
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:9]: why is this needed? shouldn't the path generated by bazel be fine?

File: c++/src/capnp/cc_capnp_library.bzl [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+_capnp_gen = rule(
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "deps": attr.label_list(providers = [capnp_provider]),
+        "data": attr.label_list(allow_files = True),
+        "outs": attr.output_list(),
+        "src_prefix": attr.string(),
+        "_capnpc": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnp_tool"),
+        "_capnpc_cxx": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-c++"),
+        "_capnpc_capnp": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-capnp"),
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:67]: maybe use toolchains instead?

File: c++/src/capnp/cc_capnp_library.bzl (hunk 3 above)
Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:9]: This is needed to generate -I directives when capnp is an external repo. Then this essentially adds "external/" prefix to paths where needed.

File: c++/src/capnp/cc_capnp_library.bzl (hunk 4 above)
Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:67]: Frankly, after setting up c++ toolchain it seems like way too much work for very little benefit.

File: c++/src/capnp/cc_capnp_library.bzl (hunk 3 above)
Comment (uhthomas) [reply c++/src/capnp/cc_capnp_library.bzl:9]: Bazel 6 removes the external/ stuff. Given this is a rule I really don't see why the file can't be passed to ctx.actions.args() and expanded automatically by Bazel.

File: c++/src/capnp/cc_capnp_library.bzl (hunk 4 above)
Comment (uhthomas) [reply c++/src/capnp/cc_capnp_library.bzl:67]: The c++ toolchain stuff is a very different and tedious beast compared to regular toolchains. A capnp toolchain would be very simple and very beneficial. I can make a mock-up tomorrow if you like.

**Comment by uhthomas** [comment]
also please add an empty `BUILD.bazel` file to `c++/`. Otherwise errors like:

```
(15:37:58) ERROR: /private/var/tmp/_bazel_thomas/796dd52d4d981101108548a9446ddca6/external/crate_index__segments-rpc-0.0.0/BUILD.bazel:5:22: no such package '@com_github_capnproto_capnproto//': The repository '@com_github_capnproto_capnproto' could not be resolved: Repository '@com_github_capnproto_capnproto' is not defined and referenced by '@crate_index__segments-rpc-0.0.0//:segments-rpc_build_script'
```

will occur.

File: c++/src/capnp/cc_capnp_library.bzl (hunk 3 above)
Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:9]: > Given this is a rule I really don't see why the file can't be passed to ctx.actions.args() and expanded automatically by Bazel.

This doesn't get files as input, but directories. E.g. src-prefix is not a file but part of the name of the file, etc. I don't see how args will help.

**Comment by mikea** [comment]
> If we're thinking of providing native rules, maybe we should also look at https://github.com/kgreenek/rules_capnproto? It seems quite fleshed out.

I feel that implementation is needlessly complicated and will definitely require lots of wrangling to adopt to this and runtime situation. I also think we prefer the original implementation for the speed.

File: c++/src/capnp/cc_capnp_library.bzl (hunk 4 above)
Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:67]: I gave it a bit of thought and I don't think toolchain is necessary in this case. These files build capnp from these sources and use it. The goal is not to provide a separate independent implementation of cc_capnp_library. I don't think we need a toolchain complexity here.

**Comment by mikea** [comment]
Comments address. PTAL @uhthomas

File: c++/src/capnp/cc_capnp_library.bzl (hunk 3 above)
Comment (uhthomas) [reply c++/src/capnp/cc_capnp_library.bzl:9]: Given

```
"_capnp_system": attr.label(default = "@capnp-cpp//src/capnp:capnp_system_library"),
```

Can't this be changed to `attr.label(allow_files = True)` and then use `ctx.files._capnp_system[0].dirname`? I just don't see how appending the workspace root makes any sense within an action.

Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:9]: src_prefix is not like this anyway.

Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:9]: but done for capnp_system

File: build/.cmake/api/v1/query/client-vscode/query.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"requests":[{"kind":"cache","version":2},{"kind":"codemodel","version":2},{"kind":"toolchains","version":1}]}
```
Comment (uhthomas) [inline build/.cmake/api/v1/query/client-vscode/query.json:1]: ?

File: c++/src/capnp/cc_capnp_library.bzl [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        name = name + "_gen",
+        srcs = srcs,
+        deps = [s + "_gen" for s in deps],
+        data = data,
+        outs = hdrs + srcs_cpp,
+        src_prefix = src_prefix,
+        visibility = visibility,
+        target_compatible_with = target_compatible_with,
+    )
+    native.cc_library(
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:111]: Why not just use [CcInfo](https://docs.bazel.build/versions/main/integrating-with-rules-cc.html#implementing-starlark-rules-that-depend-on-c-rules-andor-that-c-rules-can-depend-on)?

example from rules_go: https://github.com/bazelbuild/rules_go/blob/df02d01fb37701ff826697271220f043b364d1c1/go/private/rules/binary.bzl#L72

File: c++/src/capnp/cc_capnp_library.bzl [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+def cc_capnp_library(
+        name,
+        srcs = [],
+        data = [],
+        deps = [],
+        src_prefix = "",
+        visibility = None,
+        target_compatible_with = None,
+        **kargs):
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:83]: kwargs?

File: c++/src/capnp/cc_capnp_library.bzl [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        "data": attr.label_list(allow_files = True),
+        "outs": attr.output_list(),
+        "src_prefix": attr.string(),
+        "_capnpc": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnp_tool"),
+        "_capnpc_cxx": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-c++"),
+        "_capnpc_capnp": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-capnp"),
+        "_capnp_system": attr.label(default = "@capnp-cpp//src/capnp:capnp_system_library"),
+    },
+    output_to_genfiles = True,
+    implementation = _impl,
```
Comment (uhthomas) [inline c++/src/capnp/cc_capnp_library.bzl:72]: prefer `_capnp_gen_impl` as convention

File: c++/src/capnp/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,245 @@
+load(":cc_capnp_library.bzl", "cc_capnp_library")
```
Comment (uhthomas) [inline c++/src/capnp/BUILD.bazel:1]: prefer absolute paths (//c++/src/capnp:cc_capnp_library)

File: c++/src/capnp/cc_capnp_library.bzl (hunk 6 above)
Comment (mikea) [reply c++/src/capnp/cc_capnp_library.bzl:111]: I'm not sure what that is, it will take me some time to figure it out. Can I postpone this to a future PR?

Also what are the benefits of it? It is a cc_library anyway.

**Comment by mikea** [comment]
@uhthomas I can't get rid of remaining workspace path because of src/prefix, otherwise I think I addressed everything. Is there anything else you'd like me to change?

@kentonv could you take a quick look too?

**Review by kkoenig — APPROVED** [review]

**Comment by kkoenig** [comment]
> @uhthomas I can't get rid of remaining workspace path because of src/prefix, otherwise I think I addressed everything. Is there anything else you'd like me to change?
> 
> @kentonv could you take a quick look too?

I took a look
